analyses/cnv-chrom-plot/util/bin-coverage.R

```diff
@@ -0,0 +1,142 @@
+# Functions for calling CN statuses of genome bins
+#
+# C. Savonen for ALSF - CCDL
+#
+# 2020
+
+bp_per_bin <- function(bin_ranges, status_ranges) {
+  # Given a binned genome ranges object and another GenomicRanges object, 
+  # Return the number of bp covered per bin.
+  #
+  # Args:
+  #   bin_ranges: A binned GenomicRanges made from tileGenome. 
+  #   status_ranges:A GenomicRanges object to calculate what percent coverage of
+  #   each bin. 
+  #
+  # Returns:
+  #  a data.frame with bins x number of bp 
+  
+  # Find the portions of each copy number segment that overlap with each bin.
+  bin_overlaps <- GenomicRanges::pintersect(
+    IRanges::findOverlapPairs(
+      bin_ranges,
+      status_ranges
+    )
+  )
+  
+  # Which bins do the segs in `bin_overlaps` overlap with?
+  bin_indices <- GenomicRanges::findOverlaps(
+    bin_ranges,
+    bin_overlaps
+  )
+  
+  # Get the sum of the length of all seg portions for each bin.
+  bp_per_bin <- tapply(
+    bin_overlaps@ranges@width, # Get length of each sequence within the bin
+    bin_indices@from, # Index of which bin it overlaps
+    sum
+  ) # Add up length per bin
+  
+  # Format as data.frame with rows = bins
+  per_bin_df <- data.frame(
+    bin = as.numeric(names(bp_per_bin)),
+    bp_per_bin = as.numeric(bp_per_bin)
+  )
+  
+  # Store dummy counts if there are no ranges that are in the bins
+  if (nrow(per_bin_df) == 0) {
+    per_bin_df <- data.frame(
+      bin = as.numeric(1:length(bin_ranges)),
+      bp_per_bin = 0
+    )
+  }
```

Comment:
bp_ber_bin will not report the `0` because `findOverlaps` does not, so That's why I do this janky thing at line 49 where I make sure we have all original indices before we start joining. Then any NAs are turned to `0` before we start actually doing the fractions.